### PR TITLE
[GMS-891][Feature] preset contracts for batch minting

### PR DESCRIPTION
**Pull request by jasonzwli**

### CODE REVIEW

File: contracts/royalty-enforcement/RoyaltyEnforced.sol [hunk 1]
```diff
@@ -0,0 +1,134 @@
+//SPDX-License-Identifier: Apache 2.0
+pragma solidity ^0.8.0;
+
+// Token
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
```
Comment (EdwardJES) [inline contracts/royalty-enforcement/RoyaltyEnforced.sol:5]: unused import?

File: contracts/token/erc721/abstract/ImmutableERC721Base.sol [hunk 2]
```diff
@@ -34,8 +32,24 @@ abstract contract ImmutableERC721Base is
     /// @dev Common URIs for individual token URIs
     string public baseURI;
 
-    /// @dev the tokenId of the next NFT to be minted.
-    Counters.Counter private nextTokenId;
+    /// @dev Only MINTER_ROLE can invoke permissioned mint.
+    bytes32 public constant MINTER_ROLE = bytes32("MINTER_ROLE");
+
+    /// @dev Total amount of minted tokens to a non zero address
+    uint256 public _totalSupply = 0;
```
Comment (EdwardJES) [inline contracts/token/erc721/abstract/ImmutableERC721Base.sol:39]: nit; you don't have to initalise to 0 here, as that is the default value. `uint256 public _totalSupply;`

File: contracts/token/erc721/abstract/ImmutableERC721Base.sol [hunk 3]
```diff
@@ -145,25 +155,82 @@ abstract contract ImmutableERC721Base is
...
+            _mint(mintRequest.to, mintRequest.tokenIds[j]);
+        }
+        _totalSupply = _totalSupply + mintRequest.tokenIds.length;
+    }
+
+    /// @dev mints specified token ids to specified address
+    function _safeBatchMint(IDMint memory mintRequest) internal {
+        require(mintRequest.to != address(0), "Address is zero");
+        for (uint256 j = 0; j < mintRequest.tokenIds.length; j++) {
+            _safeMint(mintRequest.to, mintRequest.tokenIds[j]);
```
Comment (EdwardJES) [inline contracts/token/erc721/abstract/ImmutableERC721Base.sol:224]: I presume `_safeMint` also checks whether token has been burnt or not?

File: contracts/token/erc721/abstract/ImmutableERC721Base.sol [hunk 4]
```diff
@@ -145,25 +155,82 @@ abstract contract ImmutableERC721Base is
...
+    function totalSupply() public virtual view returns(uint256){
+        return _totalSupply;
+    }
+
+    ///     =====  Internal functions  =====
+
+    /// @dev mints specified token ids to specified address
+    function _batchMint(IDMint memory mintRequest) internal {
+        require(mintRequest.to != address(0), "Address is zero");
+        for (uint256 j = 0; j < mintRequest.tokenIds.length; j++) {
```
Comment (EdwardJES) [inline contracts/token/erc721/abstract/ImmutableERC721Base.sol:214]: nit; won't need to initialise to 0 here for the loop index variable.

**Review by EdwardJES — APPROVED** [review]

**Review by ipekt — APPROVED** [review]

**Pull request merged**